### DISCUSSION TITLE
Include common geofences into distribution

### DISCUSSION THREAD

**Discussion by marcogabriel** [teslamate-org/teslamate#426] [Ideas]
It would be nice to have some geofences predefined like

* all Tesla Superchargers
* all Tesla Destination Chargers
* Other Points of Interest that are more or less Tesla-specific like Tesla Service Centers, Tesla Show Rooms
* Maybe even other charging stations like IONITY, FastNed or smaller providers

The list could be maintained by a community or taken from an existing data source like

* supercharge.info -> https://forum.supercharge.info/t/downloading-supercharger-data/215/4
* goingelectric.de
* Tesla Location Finder -> https://www.tesla.com/findus

**Comment by timguy99** [comment]
I'd prefer to keep it as a user defined thing. For example I may not want all Destination chargers or some such thing. If its a separate list then it MIGHT make sense but not sure about that either.

**Comment by DrMichael** [comment]
Well, what about an import of a defined format? We could have predefined set of locations and one could construct an individual set.

**Comment by adriankumpf** [comment]
Ideally, all chargers would be directly retrievable from OpenStreetMap, right? Perhaps we should promote this topic in the OSM community? Since there are apparently already lists of all (super)chargers in a machine-readable format, an automated import into OSM should be possible? Many other projects would benefit from this as well.

**Comment by marcogabriel** [comment]
@adriankumpf you're probably right. If the data goes into OSM, it would be the best place to be.

But according to https://wiki.openstreetmap.org/wiki/Import it looks like a fairly complicated process.

**Comment by dakaix** [comment]
I've taken to exporting the Supercharger data from the OpenChargeMap API and importing it directly into Postgres, at least for my in-country stations.

https://openchargemap.org/site/develop/api
https://api.openchargemap.io/v3/poi/?output=json&maxresults=3000&operatorid=23&levelid=3&countrycode=GB   << All UK Superchargers

Ideally this wouldn't be necessary though, perhaps instead Teslamate could dynamically populate a Geo-Fence or Charging Session based on OCM data and the Lat/Long of the Charging session? Pricing is included in OCM's data for most locations (though costs are often per-country anyway!).

**Comment by ma9mwah** [comment]
I think if we are looking in to storing the pricing in the DB then we need to think about changing it over to a slowly changing dimension so that historical or even time based tariffs could be calculated. This is however quite a major change to the DB schema and would have to be looked into at quite a look of detail.

**Comment by tacotran** [comment]
> I think if we are looking in to storing the pricing in the DB then we need to think about changing it over to a slowly changing dimension so that historical or even time based tariffs could be calculated. This is however quite a major change to the DB schema and would have to be looked into at quite a look of detail.

I took a look at this because when I saw https://github.com/adriankumpf/teslamate/issues/655 because I thought it would be a cool feature despite my own charge pricing being completely static. Moving the charge pricing out of the geo-fence table would probably be easy enough, but charge cost is calculated once at the end of every charge, and not dynamically (on every grafana query) nor retroactively, Having historical data wouldn't be particularly useful for anything except fixing old charge costs.

**Comment by ma9mwah** [comment]
Ah, i assumed pricing was dynamic. If it's not then that makes it a little bit easier.

**Comment by tabascoeye** [comment]
pricing even varies for the same charging spot, depending on which card you use there.
I don't think it makes sense trying to gain that data and keep it updated. Other sites on the web are trying to do only that and it seems like a full-time job and an endless fight

**Comment by dakaix** [comment]
I agree that trying to dynamically price every charging station is an overly complex task.

However covering the Superchargers with their simplified pricing is an easy win, the only issue is with currency conversion (which could again be resolved relatively easily).

**Comment by tacotran** [comment]
> I agree that trying to dynamically price every charging station is an overly complex task.
> 
> However covering the Superchargers with their simplified pricing is an easy win, the only issue is with currency conversion (which could again be resolved relatively easily).

For time-metered superchargers, does the API tell you what tier pricing you're on? Other than the 60kW threshold, it also says if you're sharing power, which could theoretically happen beyond 60kW, up to 75kW and maybe even 125ish kW (if there are in fact V3 chargers that share power cabinets).
